cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -428,12 +524,120 @@ ArrayKernelExec ArithmeticExecFromOp(detail::GetTypeId get_id) {
   }
 }
 
+Status CastBinaryDecimalArgs(const std::string& func_name,
+                             std::vector<ValueDescr>* values) {
+  auto& left_type = (*values)[0].type;
+  auto& right_type = (*values)[1].type;
+  DCHECK(is_decimal(left_type->id()) || is_decimal(right_type->id()));
+
+  // decimal + float = float
+  if (is_floating(left_type->id())) {
+    right_type = left_type;
+    return Status::OK();
+  } else if (is_floating(right_type->id())) {
+    left_type = right_type;
+    return Status::OK();
+  }
+
+  // precision, scale of left and right args
+  int32_t p1, s1, p2, s2;
+
+  // decimal + integer = decimal
```

Comment:
integer to decimal cast is not available, this feature is not tested yet.

Reply:
well that's embarrassing. Would you add a follow up JIRA for integer->decimal casts, please?

Reply:
Sure. Created JIRA https://issues.apache.org/jira/browse/ARROW-13067.